zap/src/output/client.rs

```diff
@@ -1,260 +1,363 @@
+		let callback = casing(self.file.casing, "Callback", "callback", "callback");
+
+		self.push_line(&format!("{set_callback} = function({callback}: ({ty}) -> ())",));
+		self.indent();
+
+		self.push_line(&format!("events[{id}] = {callback}"));
+
+		self.dedent();
+		self.push_line("end,");
+	}
+
+	fn push_return_on(&mut self, ev: &EvDecl, id: usize) {
+		let ty = &ev.data;
+
+		let on = casing(self.file.casing, "On", "on", "on");
+		let callback = casing(self.file.casing, "Callback", "callback", "callback");
+
+		self.push_line(&format!("{on} = function({callback}: ({ty}) -> ())",));
+		self.indent();
+
+		self.push_line(&format!("table.insert(events[{id}], {callback})",));
+
+		self.dedent();
+		self.push_line("end,");
+	}
+
+	pub fn push_return_listen(&mut self) {
+		for (i, ev) in self
+			.file
 			.ev_decls
 			.iter()
 			.enumerate()
-			.filter(|(_, ev_decl)| ev_decl.from == EvSource::Server)
+			.filter(|(_, ev_decl)| ev_decl.from == EvSource::Client)
 		{
 			let id = i + 1;
-			let ty = &ev_decl.data;
-
-			writeln!(f, "\t{name} = {{", name = ev_decl.name,)?;
-
-			match ev_decl.call {
-				EvCall::SingleSync | EvCall::SingleAsync => {
-					writeln!(
-						f,
-						"\t\t{set_callback} = function({callback}: ({ty}) -> ())",
-						set_callback = casing(file.casing, "SetCallback", "setCallback", "set_callback"),
-						callback = casing(file.casing, "Callback", "callback", "callback"),
-					)?;
-
-					writeln!(
-						f,
-						"\t\t\tevents[{id}] = {callback}",
-						callback = casing(file.casing, "Callback", "callback", "callback")
-					)?;
-
-					writeln!(f, "\t\tend,")?;
-				}
-
-				EvCall::ManySync | EvCall::ManyAsync => {
-					writeln!(
-						f,
-						"{on} = function({callback}: ({ty}) -> ())",
-						on = casing(file.casing, "On", "on", "on"),
-						callback = casing(file.casing, "Callback", "callback", "callback"),
-					)?;
-
-					writeln!(
-						f,
-						"\t\t\ttable.insert(events[{id}], {callback})",
-						callback = casing(file.casing, "Callback", "callback", "callback")
-					)?;
-
-					writeln!(f, "\t\tend,")?;
-				}
+
+			self.push_line(&format!("{name} = {{", name = ev.name));
+			self.indent();
+
+			match ev.call {
+				EvCall::SingleSync | EvCall::SingleAsync => self.push_return_setcallback(ev, id),
+				EvCall::ManySync | EvCall::ManyAsync => self.push_return_on(ev, id),
 			}
 
-			writeln!(f, "\t}},")?;
+			self.dedent();
+			self.push_line("}},");
 		}
+	}
+
+	pub fn push_return(&mut self) {
+		self.push_line("return {");
+		self.indent();
+
+		self.push_return_outgoing();
+		self.push_return_listen();
+
+		self.dedent();
+		self.push_line("}");
+	}
 
-		writeln!(f, "}}")?;
+	pub fn output(&mut self) {
+		self.push_file_header();
+
+		self.push(include_str!("client.luau"));
+
+		self.push_tydecls();
+
+		self.push_reliable();
+		self.push_unreliable();
+
+		self.push_callback_lists();
+
+		self.push_return();
+	}
 
-		Ok(())
+	pub fn into_inner(self) -> String {
+		self.buff
 	}
```

Comment:
Do you really need this function?

Reply:
Not really, we can remove it and put it into `output`